File: Day-10/python/das_/solution.py
import itertools

def solve(row, column, n=0):
    if not (0 <= row < len(grid) and 0 <= column < len(grid[0])):
        return []

    if grid[row][column] != n:
        return []

    if n == 9:
        return [(row, column)]

    return (solve(row + 1, column, n + 1) +
            solve(row, column + 1, n + 1) +
            solve(row - 1, column, n + 1) +
            solve(row, column - 1, n + 1))

grid = [list(map(int, line.strip())) for line in open("input.txt")]
paths = [solve(*pos) for pos in itertools.product(range(len(grid)), range(len(grid[0])))]
print("Part 1:", sum(len(set(path)) for path in paths))
print("Part 2:", sum(len(path) for path in paths))

File: Day-10/haskell/das_/Solution.hs
import Data.Char (digitToInt)
import Data.List (group, sort)
import Text.Printf (printf)

solve :: [[Int]] -> Int -> Int -> Int -> [(Int, Int)]
solve grid row column n
    | not (0 <= row && row < length grid && 0 <= column && column < length (head grid)) = []
    | grid !! row !! column /= n = []
    | n == 9 = [(row, column)]
    | otherwise = concat [solve grid (row + fst d) (column + snd d) (n + 1) | d <- [(0, 1), (1, 0), (0, -1), (-1, 0)]]

main :: IO ()
main = do
    content <- readFile "input.txt"
    let grid = [map digitToInt xs | xs <- lines content]
    let coordinates = [(row, column) | row <- [0..(length grid - 1)], column <- [0..(length (head grid) - 1)]]
    let paths = map (\pos -> uncurry (solve grid) pos 0) coordinates
    printf "Part 1: %d\n" $ sum $ map (\pos -> length $ map head . group . sort $ pos) paths
    printf "Part 2: %d\n" $ sum $ map length paths
